teps/XXXX-step-workspaces.md

```diff
@@ -0,0 +1,254 @@
+---
+title: step-and-sidecar-workspaces
+authors:
+  - "@sbwsg"
+creation-date: 2020-10-02
+last-updated: 2020-10-02
+status: proposed
+---
+
+-->
+# TEP-XXXX: Step and Sidecar Workspaces
+
+<!--
+Ensure the TOC is wrapped with
+  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
+tags, and then generate with `hack/update-toc.sh`.
+-->
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Add <code>workspaces</code> to <code>Steps</code>](#add--to-)
+  - [Add <code>workspaces</code> to <code>Sidecars</code>](#add--to--1)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+    - [Story 4](#story-4)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [Use an explicit volumeMount instead](#use-an-explicit-volumemount-instead)
+    - [Advantages](#advantages)
+    - [Drawbacks](#drawbacks-1)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+## Motivation
+
+### Goals
+
+- Provide a mechanism to limit the exposure of sensitive workspaces to only those Steps in a Task
+that actually require access to them.
+- Provide explicit access to Task workspaces from sidecars without using `volumeMounts`.
+
+### Non-Goals
+
+- Validation of the contents of a WorkspaceBinding
+- Validation of the configuration of a WorkspaceBinding
+- Overriding configuration of a WorkspaceBinding
+
+## Requirements
+
+- A Task author should be able to limit access to a `Workspace` to only those `Steps` that actually
+require the contents of that `Workspace`. By doing so they should be able to limit the running code
+that has access to those contents as well.
+- A Task author should be able to use a Workspace from a `Sidecar`.
+
+## Proposal
+
+### Add `workspaces` to `Steps`
+
+1. Add a `workspaces` list to `Steps`.
+2. Allow `workspaces` from the `Task` to be explicitly named in that list like this:
+
+    ```yaml
+    workspaces:
+    - name: my-sensitive-workspace
+    steps:
+    - name: foo
+      workspaces:
+      - name: my-sensitive-workspace
+    ```
+3. A `workspace` named in a `Step` is not given to other `Steps` unless they too have the
+`workspace` in their own `workspaces` list.
+
+Example YAML:
+
+```yaml
+# task spec
+spec:
+  workspaces:
+  - name: git-ssh-credentials
+    mountPath: /root/.ssh
```

Comment:
Do we want to be able to have different `mountPath` per steps ? Imagine we have a workspace in a Task that we use in 2 steps (out of 6), but want/need this workspace to be mounted at different path ?

Reply:
I could see utility in overriding the `mountPath` on a per-Step basis, yeah.  I'll spend some time thinking of the use-cases it might be desirable for.

Do you think `mountPath` should also be a part of this TEP or tackled separately?  If we decide to move ahead with #235 in some form then there may be other fields that are desirable per-Step as well.